### PR TITLE
[bug] [opt] Fix some bugs when deal with real function

### PR DESCRIPTION
**Pull request by mzmzm**
Related issue = #602
Fix some passes to deal with real function.

### CODE REVIEW

File: taichi/program/kernel.h [hunk 1]
```diff
@@ -19,6 +19,7 @@ class TI_DLL_EXPORT Kernel : public Callable {
 
   bool is_accessor{false};
   bool is_evaluator{false};
+  bool has_real_func{false};
```
Comment (lin-hitonami) [inline taichi/program/kernel.h:22]: Is `has_real_func` still needed? I can't find any usage of it in this PR.

Comment (mzmzm) [reply taichi/program/kernel.h:22]: in [ir_node_extended_impl.cpp#30](https://github.com/taichi-dev/taichi/pull/4568/files#diff-ba3a420d15b12577ed1475815e2523183ee9afd8921445aadab7f4cc67a073b9R30)

Comment (lin-hitonami) [reply taichi/program/kernel.h:22]: I mean is all those `has_real_func` and `set_has_real_func` functions and variables still needed?

Comment (mzmzm) [reply taichi/program/kernel.h:22]: I use `Funcallstmt` instead of the check, let me remove it.

**Review by lin-hitonami — APPROVED** [review]
LGTM! Thanks!

**Pull request merged**